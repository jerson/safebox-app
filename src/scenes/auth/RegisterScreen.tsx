import React, { useState, useRef } from 'react';
import {
  StyleSheet,
  View,
  ViewStyle,
  ScrollView,
  StatusBar
} from 'react-native';
import { SafeAreaView, NavigationScreenProp } from 'react-navigation';
import Colors from '../../modules/constants/Colors';
import HeaderLanding from '../../components/ui/HeaderLanding';
import TextInput, { TextInputRef } from '../../components/ui/TextInput';
import Button from '../../components/ui/Button';
import Container from '../../components/ui/Container';
import Content from '../../components/ui/Content';
import Size from '../../modules/dimensions/Size';
import useTextInput from '../../components/hooks/useTextInput';
import Client from '../../services/Client';
import { RegisterRequest } from '../../proto/services_pb';
import OpenPGP, { KeyOptions } from 'react-native-fast-openpgp';
import Config from '../../Config';
import Session from '../../services/Session';
import Strings from '../../modules/format/Strings';
import AlertMessage from '../../components/ui/AlertMessage';
import ButtonLink from '../../components/ui/ButtonLink';
import useAnimatedState from '../../components/hooks/useAnimatedState';

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.grey1
  } as ViewStyle,
  scrollView: {
    flex: 1
  } as ViewStyle,
  safeArea: {
    flex: 1
  } as ViewStyle,
  headerLanding: {
    marginTop: 20,
    marginBottom: 20
  } as ViewStyle,
  textInput: {} as ViewStyle,
  textInputContainer: {
    marginBottom: 0
  } as ViewStyle,
  form: {
    width: 280,
    marginBottom: 60,
    alignSelf: 'center'
  } as ViewStyle,
  button: {
    marginTop: 20
  } as ViewStyle,
  buttonLink: {
    marginTop: 40
  } as ViewStyle
});

interface Params {}
interface Props {
  navigation: NavigationScreenProp<Params>;
}

function RegisterScreen({ navigation }: Props) {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useAnimatedState('');

  const usernameRef = useRef<TextInputRef>(null);
  const passwordRef = useRef<TextInputRef>(null);
  const repeatPasswordRef = useRef<TextInputRef>(null);

  const [username, usernameProps] = useTextInput('');
  const [password, passwordProps] = useTextInput('');
  const [repeatPassword, repeatPasswordProps] = useTextInput('');

  const isValid = () => {
    const isValid = [!!username, !!password, !!repeatPassword].every(
      value => value
    );
    if (!isValid) {
      setError('Complete missing fields');
      return false;
    }
    if (password !== repeatPassword) {
      setError('Passwords must match');
      return false;
    }
    return isValid;
  };

  const generateKeyPair = () => {
    return OpenPGP.generate({
      passphrase: password,
      name: username,
      keyOptions: Config.settings.keyOptions as KeyOptions
    });
  };

  const submit = async () => {
    try {
      const keyPair = await generateKeyPair();

      const request = new RegisterRequest();
      request.setUsername(username);
      request.setPrivatekey(keyPair.privateKey);
      request.setPublickey(keyPair.publicKey);

      const response = await Client.register(request);

      Session.login(response.getAccesstoken());
      navigation.navigate('Accounts');
    } catch (e) {
      const message = Strings.getError(e);
      setError(message);
    }
  };
  const tryToSubmit = () => {
    if (!isValid()) {
      return;
    }

    setIsLoading(true);
    requestAnimationFrame(async () => {
      await submit();
      setIsLoading(false);
    });
  };

  const goToLogin = () => {
    navigation.goBack();
  };

  return (
    <Container style={styles.container}>
      <StatusBar
        animated
        barStyle={'dark-content'}
        backgroundColor={Colors.grey2}
      />
      <ScrollView
        keyboardShouldPersistTaps={'handled'}
        contentContainerStyle={{
          minHeight: Size.getVisibleHeight()
        }}
        style={styles.scrollView}
      >
        <SafeAreaView style={styles.safeArea}>
          <Content center>
            <HeaderLanding
              titleStyle={{ color: Colors.primary }}
              subtitleStyle={{ color: Colors.primary }}
              style={styles.headerLanding}
            />
            <View style={styles.form}>
              {!!error && <AlertMessage message={error} />}
              <TextInput
                icon={'user'}
                placeholder={'Username'}
                keyboardType={'default'}
                autoCapitalize={'none'}
                autoCorrect={false}
                autoCompleteType={'username'}
                returnKeyType={'next'}
                containerStyle={styles.textInputContainer}
                style={styles.textInput}
                ref={usernameRef}
                onSubmitEditing={() => {
                  passwordRef.current && passwordRef.current.focus();
                }}
                {...usernameProps}
              />
              <TextInput
                icon={'lock'}
                placeholder={'Password'}
                secureTextEntry
                keyboardType={'default'}
                autoCapitalize={'none'}
                autoCorrect={false}
                autoCompleteType={'password'}
                returnKeyType={'next'}
                containerStyle={styles.textInputContainer}
                style={styles.textInput}
                ref={passwordRef}
                onSubmitEditing={() => {
                  repeatPasswordRef.current &&
                    repeatPasswordRef.current.focus();
                }}
                {...passwordProps}
              />
              <TextInput
                icon={'lock'}
                placeholder={'Repeat password'}
                secureTextEntry
                keyboardType={'default'}
                autoCapitalize={'none'}
                autoCorrect={false}
                autoCompleteType={'password'}
                returnKeyType={'done'}
                containerStyle={styles.textInputContainer}
                style={styles.textInput}
                ref={repeatPasswordRef}
                blurOnSubmit
                {...repeatPasswordProps}
              />

              <Button
                isLoading={isLoading}
                style={styles.button}
                typeColor={'primaryLight'}
                title={'Create account'}
                onPress={tryToSubmit}
              />

              <ButtonLink
                style={styles.buttonLink}
                title={'Back to Sign In'}
                onPress={goToLogin}
              />
            </View>
          </Content>
        </SafeAreaView>
      </ScrollView>
    </Container>
  );
}

RegisterScreen.navigationOptions = {
  headerLeft: null,
  headerRight: null,
  title: null
};

export default RegisterScreen;
