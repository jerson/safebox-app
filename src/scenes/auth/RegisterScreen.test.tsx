import React from 'react';
import * as renderer from 'react-test-renderer';
import RegisterScreen from './RegisterScreen';

const props = {};
test('renders without crashing', () => {
  const snapshot = renderer.create(<RegisterScreen {...props} />).toJSON();
  expect(snapshot).toBeTruthy();
});

test('render just views', () => {
  const snapshot = renderer.create(<RegisterScreen {...props} />).toJSON();
  expect(snapshot).toMatchSnapshot();
});
