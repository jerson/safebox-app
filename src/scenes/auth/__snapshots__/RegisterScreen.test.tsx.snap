// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render just views 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fafdff",
        "flex": 1,
      },
      Object {
        "backgroundColor": "#fafdff",
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "minHeight": 1314,
      }
    }
    keyboardShouldPersistTaps="handled"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        onLayout={[Function]}
        pointerEvents="box-none"
        style={
          Object {
            "flex": 1,
            "paddingBottom": 0,
            "paddingLeft": 0,
            "paddingRight": 0,
            "paddingTop": 20,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "paddingHorizontal": 20,
              },
              Object {
                "justifyContent": "center",
              },
              Object {
                "flex": 1,
                "paddingTop": 20,
              },
              undefined,
            ]
          }
        >
          <View
            onLayout={[Function]}
            style={
              Object {
                "paddingBottom": 0,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "justifyContent": "center",
                    "overflow": "visible",
                  },
                  false,
                  Object {
                    "marginBottom": 20,
                    "marginTop": 20,
                  },
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignSelf": "center",
                    "backgroundColor": "#38414e",
                    "borderColor": "#c2d0e7",
                    "borderRadius": 22,
                    "borderWidth": 1,
                    "overflow": "hidden",
                  }
                }
              >
                <Image
                  resizeMode="stretch"
                  source={
                    Object {
                      "testUri": "../../../src/assets/images/logo.png",
                    }
                  }
                  style={
                    Object {
                      "height": 90,
                      "width": 90,
                    }
                  }
                />
              </View>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Nunito",
                      "fontSize": undefined,
                      "fontStyle": "normal",
                      "fontWeight": "400",
                    },
                    Object {},
                    Array [
                      Object {
                        "color": "#FFFFFF",
                        "fontSize": 35,
                      },
                      Object {
                        "color": "#4531cd",
                      },
                    ],
                  ]
                }
              >
                Safe
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Nunito",
                        "fontSize": undefined,
                        "fontStyle": "normal",
                        "fontWeight": "700",
                      },
                      Object {},
                      undefined,
                    ]
                  }
                >
                  Box
                </Text>
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "fontFamily": "Nunito",
                      "fontSize": undefined,
                      "fontStyle": "normal",
                      "fontWeight": "300",
                    },
                    Object {},
                    Array [
                      Object {
                        "color": "#f2f4f9",
                        "fontSize": 14,
                      },
                      Object {
                        "color": "#4531cd",
                      },
                    ],
                  ]
                }
              >
                secrets always safe
              </Text>
            </View>
            <View
              style={
                Object {
                  "alignSelf": "center",
                  "marginBottom": 60,
                  "width": 280,
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "column",
                      "marginBottom": 20,
                      "marginTop": 10,
                      "position": "relative",
                    },
                    Object {
                      "marginBottom": 0,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "paddingBottom": 4,
                      },
                      undefined,
                    ]
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "position": "relative",
                      },
                      undefined,
                    ]
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    autoCapitalize="none"
                    autoCompleteType="username"
                    autoCorrect={false}
                    editable={true}
                    keyboardType="default"
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="Username"
                    placeholderTextColor="#65748b"
                    rejectResponderTermination={true}
                    returnKeyType="next"
                    selectionColor="#65748b"
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                          "borderColor": "#f2f4f9",
                          "borderRadius": 10,
                          "borderWidth": 1,
                          "color": "#38414e",
                          "fontFamily": "Nunito",
                          "fontSize": 15,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "paddingBottom": 10,
                          "paddingLeft": 10,
                          "paddingRight": 8,
                          "paddingTop": 10,
                          "textDecorationLine": "none",
                        },
                        Object {
                          "paddingLeft": 40,
                        },
                        false,
                        false,
                        Object {
                          "paddingRight": 40,
                        },
                        Object {},
                      ]
                    }
                    underlineColorAndroid="transparent"
                    value=""
                  />
                  <View
                    style={
                      Object {
                        "bottom": 1,
                        "justifyContent": "center",
                        "left": 12,
                        "position": "absolute",
                        "top": 0,
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": undefined,
                            "fontSize": 12,
                          },
                          Array [
                            Object {
                              "color": "#65748b",
                              "fontSize": 20,
                            },
                            false,
                            false,
                            undefined,
                          ],
                          Object {
                            "fontFamily": "Feather",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      ï‡¶
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "position": "absolute",
                          "right": -60,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "column",
                      "marginBottom": 20,
                      "marginTop": 10,
                      "position": "relative",
                    },
                    Object {
                      "marginBottom": 0,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "paddingBottom": 4,
                      },
                      undefined,
                    ]
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "position": "relative",
                      },
                      undefined,
                    ]
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    autoCapitalize="none"
                    autoCompleteType="password"
                    autoCorrect={false}
                    editable={true}
                    keyboardType="default"
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="Password"
                    placeholderTextColor="#65748b"
                    rejectResponderTermination={true}
                    returnKeyType="next"
                    secureTextEntry={true}
                    selectionColor="#65748b"
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                          "borderColor": "#f2f4f9",
                          "borderRadius": 10,
                          "borderWidth": 1,
                          "color": "#38414e",
                          "fontFamily": "Nunito",
                          "fontSize": 15,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "paddingBottom": 10,
                          "paddingLeft": 10,
                          "paddingRight": 8,
                          "paddingTop": 10,
                          "textDecorationLine": "none",
                        },
                        Object {
                          "paddingLeft": 40,
                        },
                        false,
                        false,
                        Object {
                          "paddingRight": 40,
                        },
                        Object {},
                      ]
                    }
                    underlineColorAndroid="transparent"
                    value=""
                  />
                  <View
                    style={
                      Object {
                        "bottom": 1,
                        "justifyContent": "center",
                        "left": 12,
                        "position": "absolute",
                        "top": 0,
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": undefined,
                            "fontSize": 12,
                          },
                          Array [
                            Object {
                              "color": "#65748b",
                              "fontSize": 20,
                            },
                            false,
                            false,
                            undefined,
                          ],
                          Object {
                            "fontFamily": "Feather",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      ï†‰
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "position": "absolute",
                          "right": -60,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
              </View>
              <View
                style={
                  Array [
                    Object {
                      "flexDirection": "column",
                      "marginBottom": 20,
                      "marginTop": 10,
                      "position": "relative",
                    },
                    Object {
                      "marginBottom": 0,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "paddingBottom": 4,
                      },
                      undefined,
                    ]
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "position": "relative",
                      },
                      undefined,
                    ]
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    autoCapitalize="none"
                    autoCompleteType="password"
                    autoCorrect={false}
                    blurOnSubmit={true}
                    editable={true}
                    keyboardType="default"
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    onSubmitEditing={[Function]}
                    placeholder="Repeat password"
                    placeholderTextColor="#65748b"
                    rejectResponderTermination={true}
                    returnKeyType="done"
                    secureTextEntry={true}
                    selectionColor="#65748b"
                    style={
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                          "borderColor": "#f2f4f9",
                          "borderRadius": 10,
                          "borderWidth": 1,
                          "color": "#38414e",
                          "fontFamily": "Nunito",
                          "fontSize": 15,
                          "fontStyle": "normal",
                          "fontWeight": "400",
                          "paddingBottom": 10,
                          "paddingLeft": 10,
                          "paddingRight": 8,
                          "paddingTop": 10,
                          "textDecorationLine": "none",
                        },
                        Object {
                          "paddingLeft": 40,
                        },
                        false,
                        false,
                        Object {
                          "paddingRight": 40,
                        },
                        Object {},
                      ]
                    }
                    underlineColorAndroid="transparent"
                    value=""
                  />
                  <View
                    style={
                      Object {
                        "bottom": 1,
                        "justifyContent": "center",
                        "left": 12,
                        "position": "absolute",
                        "top": 0,
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": undefined,
                            "fontSize": 12,
                          },
                          Array [
                            Object {
                              "color": "#65748b",
                              "fontSize": 20,
                            },
                            false,
                            false,
                            undefined,
                          ],
                          Object {
                            "fontFamily": "Feather",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      ï†‰
                    </Text>
                  </View>
                  <View
                    style={
                      Array [
                        Object {
                          "position": "absolute",
                          "right": -60,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
              </View>
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "backgroundColor": "#9066f5",
                    "borderColor": "#9066f5",
                    "borderRadius": 10,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "margin": 0,
                    "marginTop": 20,
                    "opacity": 1,
                    "padding": 12,
                    "zIndex": 1,
                  }
                }
                testID="button_touchable"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Nunito",
                        "fontSize": undefined,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                      },
                      Object {},
                      Array [
                        Object {
                          "color": "#FFFFFF",
                          "fontSize": 14,
                          "textAlign": "center",
                        },
                        Object {
                          "color": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  Create account
                </Text>
              </View>
              <View
                accessible={true}
                focusable={true}
                isTVSelectable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignSelf": "center",
                    "marginTop": 40,
                    "opacity": 1,
                    "padding": 6,
                  }
                }
                testID="touchable"
              >
                <Text
                  style={
                    Array [
                      Object {
                        "fontFamily": "Nunito",
                        "fontSize": undefined,
                        "fontStyle": "normal",
                        "fontWeight": "400",
                      },
                      Object {},
                      Array [
                        Object {
                          "color": "#65748b",
                          "fontSize": 13,
                        },
                        undefined,
                        undefined,
                        undefined,
                      ],
                    ]
                  }
                >
                  Back to Sign In
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
